src/Mollie/WC/Payment/Payment.php

```diff
@@ -285,15 +285,24 @@ public function onWebhookPaid( WC_Order $order, $payment, $payment_method_title
 	public function onWebhookCanceled( WC_Order $order, $payment, $payment_method_title ) {
 
 		// Get order ID in the correct way depending on WooCommerce version
-		if ( version_compare( WC_VERSION, '3.0', '<' ) ) {
-			$order_id = $order->id;
-		} else {
-			$order_id = $order->get_id();
-		}
+        $order_id = wooCommerceOrderId($order);
 
 		// Add messages to log
-		Mollie_WC_Plugin::debug( __METHOD__ . ' called for payment ' . $order_id );
+		debug( __METHOD__ . ' called for payment ' . $order_id );
+
+		// if the status is Completed|Refunded|Cancelled  DONT change the status to cancelled
+        $data_helper = getDataHelper();
+        $isCompleted = $data_helper->hasOrderStatus($order, 'completed');
+        $isRefunded = $data_helper->hasOrderStatus($order, 'refunded');
+        $isCancelled = $data_helper->hasOrderStatus($order, 'cancelled');
+
+        if ( $isCompleted || $isRefunded || $isCancelled){
+            debug( __METHOD__ . ' called for payment ' . $order_id . ' has final status. Nothing to be done' );
+
+            return;
+        }
```

Comment:
You can make the code more Dry if you put the logic into a method of the parent class, so you can reuse it in Order and Payment class and define just once.

Alternatively a trait would be also ok, since the parent class seems to have concerns on Payment Object only but double check it.